torch/csrc/autograd/autograd_not_implemented_fallback.cpp

```diff
@@ -45,8 +45,170 @@ void _foreach_tensor(
+        deleteNode);
+    grad_fn->set_next_edges(
+        collect_next_edges(tensors_requiring_grad_on_stack));
+  }
+
+  op.redispatchBoxed(dispatch_keys & c10::after_autograd_keyset, stack);
+
+  if (any_input_requires_grad) {
+    // NB: if the operator mutates any inputs in-place and does not return them
+    // as outputs, we are unable to lazily raise a warning. This is OK because
+    // we don't expect many existing operators to do this because of the amount
+    // of technical expertise necessary (you would need to manually register an
+    // autograd kernel without using autograd.Function)
+    _foreach_tensor(
+        [&](size_t _, size_t idx_ret, const at::Tensor& t) {
+          if (!isDifferentiableType(t.scalar_type())) {
+            return;
+          }
+          const bool is_mutable_output =
+              schema.is_aliasing({c10::SchemaArgType::output, idx_ret}) &&
+              schema.is_mutable({c10::SchemaArgType::output, idx_ret});
+
+          // If the post-autograd implementation returns Tensors that require
+          // grad, then we install a hook that will warn during the backwards.
+          //
+          // NB: If the operation is inplace and the inputs were views,
+          // it is possible that the history was rebased and the hook will
+          // not warn in all places where it should. That is, the following
+          // won't warn:
+          // >>> x = torch.randn(3, 3, requires_grad=True)
+          // >>> z = x.clone()
+          // >>> w = z[0]
+          // >>> k = w[0]
+          // >>> y = op(k)
+          // >>> torch.autograd.grad(z.sum(), w)
+          if (t.requires_grad()) {
+            t.register_hook([op_name](const at::Tensor& grad) {
+              warnAutogradNotImplemented(op_name);
+              return grad;
+            });
+            // If history is rebased, then we will attempt to warn
+            // on the view's base. This will catch most cases (because
+            // users typically call .backward() and backprop through
+            // the entire program).
+            if (t.is_view() && is_mutable_output) {
+              // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+              const_cast<at::TensorBase&>(t._base()).register_hook(
+                  [op_name](const at::TensorBase& grad) {
+                    warnAutogradNotImplemented(op_name);
+                    return grad;
+                  });
+            }
+            return;
+          }
+
+          // If the post-autograd implementation returns any Tensors that
+          // don't require grad, then we install the WarnNotImplemented grad_fn.
+          // This grad_fn warns in backward and returns undefined tensor
+          // gradients.
+          if (is_mutable_output) {
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+            rebase_history(const_cast<at::Tensor&>(t), grad_fn);
```

Comment:
I suppose we would reach here if the user passed in a view that doesn't require grad as input and return that input as-is. (User may have also done inplace ops, and if those operations involved any other inputs that requires grad it would've necessarily been done in no-grad).

I think we should just not do anything in this case, rather than rebase_history. If we're doing rebase_history, we're making the view and base which previously did not require grad, now require grad. I can see that being potentially problematic if the user uses the base/view in subsequent computation not expecting that it now requires grad and saving some saved tensors unnecessarily.

Reply:
Note that the user would have also marked that particular output as being written to in the schema, e.g. `(Tensor(a!) x) -> Tensor(a!)`. We can avoid the rebase_history to be on the safe side (to be less BC-breaking), it just comes at a tradeoff of not being able to catch situations where the user does do an in-place operation on an input view and we aren't able to warn them that they didn't add an autograd kernel

Reply:
If they had other returns, then they should still get a warning when doing backward. I guess we would no longer have a warning for the case where the user's only return was the in-place one. 
The only case where that output would naturally still be requires_grad=False is any computation involving other inputs that require grad were done in no-grad mode - in that case it seems that the users intention is to have the kernel be non-differentiable.

Reply:
Actually the real reason to not do this is because rebase_history on a tensor that doesn't require grad is undefined behavior. 
You end up having a graph that doesn't even end with a leaf node!

Reply:
I think we're missing an assert somewhere

Reply:
> Actually the real reason to not do this is because rebase_history on a tensor that doesn't require grad should be considered undefined behavior.
You end up having a graph that doesn't even end with a leaf node!

Fair enough. When we get to this point the Tensor doesn't require grad :P.
So, I'll delete the rebase_history and we will live with some code that doesn't catch all of the cases.

Reply:
> Actually the real reason to not do this is because rebase_history on a tensor that doesn't require grad should be considered undefined behavior.
You end up having a graph that doesn't even end with a leaf node!

I looked into this a bit more, and I realize what the problem is now.

So both of the following are fine without using custom ops. So clearly it is supposed to work. (sorry I should've checked these cases earlier)
```
# Rebasing on tensor that does not require grad
a = torch.tensor(1.)
b = torch.tensor(1., requires_grad=True)
a += b

# Rebasing on a view that does not grad, and whose base does not require grad
a = torch.tensor([1.])[:]
b = torch.tensor(1., requires_grad=True)
a += b
print(a._base.grad_fn.next_functions)
```

The key difference between what this fallback does and what a VariableType kernel normally does it that normally a VariableType kernel collects next edges from ALL of the inputs, including the ones that do not require grad, versus in this PR, we only collect next edges from inputs that do require grad. Normally this is fine, but the difference matters in the specific case where we do inplace over view, where we need to create a CopySlices node.

Taking a look at the CopySlices node, it notably skips over the first edge in the grad_fn its supposed to wrap, i.e. the one corresponding to the tensor being inplaced. This means that it is actually necessary to have collected the edge from the first input even though it does not require grad as to accomodate this skip.

Reply:
Resolved offline:
- if an output is mutated according to the schema and doesn't require grad, we will not do anything (rebase or set history).
- We can improve this case in the future, but right now we are OK with the tradeoff.